rpcs3/Emu/system_progress.cpp

```diff
@@ -24,8 +24,20 @@ namespace rsx::overlays
 {
 	class progress_dialog : public message_dialog
 	{
+		shared_mutex m_mutex;
 	public:
 		using message_dialog::message_dialog;
+
+		compiled_resource get_compiled() override {
+			reader_lock lock(m_mutex);
+			return message_dialog::get_compiled();
+		};
+
+		void set_text(const std::string& text) override {
```

Comment:
A minor improvement to my earlier suggestion, let's use unique_lock + std::try_to_lock. If we cannot acquire the lock, don't bother updating the text at all. It serves no purpose to stall waiting for an update to a progress dialog anyway.

Reply:
OK, done. In the previous commit I missed `progress_bar_set_message`.

Reply:
In my opinion the get compiled should be skipped instead of the setText.
It's more important to get the new data in than to compile immediately

Reply:
The renderer is passively reading the object when we're making the change which is the root of the problem. Stalling to change text actually slows down the worker thread writing to the object which is undesirable.
Also, if you randomly skip gathering the renderable representation of the object, what you get is visual flickering which is way worse.

Reply:
what flickering?
worst case it will show the old text again for another frame.

Reply:
Skipping the set_text is actually a lot worse. Remember, we use this to show certain things, like "Stopping emulation, please wait"

Reply:
Also, it may result in showing the wrong text for one of the bars for a long time...

Reply:
> what flickering? worst case it will show the old text again for another frame.

If we cannot safely get the compiled output, we'll return empty draw command which means the dialog is not visible for a frame. User-interfaces are higher-level and do not cache unlike widgets. The flip has already completed by the time we start drawing the dialogs and we have a blank frame (or ingame frame) already in the pipe so its not like you'll just see the previous contents.

Reply:
> worst case it will show the old text again for another frame.

Incorrect assumption, see above.

Reply:
> Skipping the set_text is actually a lot worse. Remember, we use this to show certain things, like "Stopping emulation, please wait"

If it's that serious, maybe we need to rethink how we change the text for progress dialogs. We can return a bool on fail to change text and force critical messages to retry until success.

Reply:
> Also, it may result in showing the wrong text for one of the bars for a long time...

For one frame.

Reply:
Another option is to have 2 versions of the function to alter the text. One to be used strictly for non-critical things like "Compiling module X of Y" and another that fully stalls for serious messages.

Reply:
This is what I had in mind:
https://github.com/RPCS3/rpcs3/compare/master...Megamouse:rpcs3:bootfix